x-pack/plugins/reporting/server/export_types/common/pdf/pdfmaker.ts

```diff
@@ -124,37 +143,69 @@ export class PdfMaker {
     this._title = title;
   }
 
-  generate() {
-    const docTemplate = _.assign(
-      getTemplate(this._layout, this._logo, this._title, tableBorderWidth, assetPath),
-      {
-        content: this._content,
-      }
-    );
-    this._pdfDoc = this._printer.createPdfKitDocument(docTemplate, getDocOptions(tableBorderWidth));
-    return this;
+  private getWorkerData(): GeneratePdfRequest['data'] {
+    return {
+      layout: {
+        hasFooter: this._layout.hasFooter,
+        hasHeader: this._layout.hasHeader,
+        orientation: this._layout.getPdfPageOrientation(),
+        useReportingBranding: this._layout.useReportingBranding,
+        pageSize: this._layout.getPdfPageSize({
+          pageMarginTop,
+          pageMarginBottom,
+          pageMarginWidth,
+          tableBorderWidth,
+          headingHeight,
+          subheadingHeight,
+        }),
+      },
+      title: this._title,
+      logo: this._logo,
+      content: this._content as unknown as SerializableRecord[],
+    };
   }
 
-  getBuffer(): Promise<Buffer | null> {
-    return new Promise((resolve, reject) => {
-      if (!this._pdfDoc) {
-        throw new Error(
-          i18n.translate(
-            'xpack.reporting.exportTypes.printablePdf.documentStreamIsNotgeneratedErrorMessage',
-            {
-              defaultMessage: 'Document stream has not been generated',
-            }
-          )
-        );
-      }
-
-      const concatStream = concat(function (pdfBuffer: Buffer) {
-        resolve(pdfBuffer);
-      });
+  public async generate(): Promise<Uint8Array> {
+    if (this.workerPort) throw new Error('PDF generation already in progress!');
+    const { port1, port2 } = new MessageChannel();
+    this.workerPort = port1;
 
-      this._pdfDoc.on('error', reject);
-      this._pdfDoc.pipe(concatStream);
-      this._pdfDoc.end();
+    const workerInstance = getWorkerInstance({
```

Comment:
Here we have a long-lived worker. I'm thinking maybe it is better to spin a new one every time? It would consume a bit of resources to spin in up, but that would be negligible for us. But we would gain:

- It would be potentially more reliable, as we create a new worker each time, instead of reusing existing one, where it is not clear in what state the existing worker is.
- We would release the memory, when reporting is not used.

Reply:
When the task that runs on the worker thread is relatively short compared to the time it takes to start up the worker then throughput suffers a lot when spinning up a new worker every time. That's why thread-based web servers all use a pool of threads to handle incoming requests.

We could benchmark it to see how badly it would affect the throughput of reporting specifically (it would mostly depend on the PDF size). However, medium term we will expose a worker pool from core for any cpu-intensive tasks which could block the event loop. So having a long-lived worker will bring this implementation in line with that.

> It would be potentially more reliable, as we create a new worker each time, instead of reusing existing one, where it is not clear in what state the existing worker is.

We are already restarting workers on any errors. If a worker has a memory leak then restarting it could help, but pdf generation is mature code so I don't think it's a high risk that this code would start negatively impacting a worker's state if it didn't do so before.

> We would release the memory, when reporting is not used.

Although garbage collection can take some time, a worker will also release any unused memory so it would just be the overhead of the process/stack that remains in memory.

(assuming a docker container such as on cloud) Keeping memory spikes below the available memory is more important than keeping the average low. By limiting the max-old-space-size of the main and worker process we can prevent Kibana from getting OOM'd. Lowering the average memory consumption of the worker thread would only be useful if we allowed the main Kibana process to use the freed memory, but this would imply we under allocated memory (e.g. kibana heap + worker heap > available memory) which would be dangerous.

Reply:
> And we don't have a requirement to optimize the throughput (as the throughput is few requests per day); instead we want to optimize for reliability.

I really want to emphasis this statement that @vadimkibana wrote. We will optimize throughput in a later stage, when we have an external service available to execute the report jobs.